### ISSUE TITLE
Gradle 8.1 and 8.1.1 alter Gradle application bytecode so that it fails anti-tamper check.

### ISSUE THREAD

**Issue by zelix0** [gradle/gradle#25159]
### Expected Behavior

Our application automatically checks that its classes have not been altered as part of an anti-tamper check. If the bytecode is altered or if the sequence of class initialization is altered then it will fail to work properly.  The expected behavior is that the anti-tamper self checks should cause to problems.

### Current Behavior (optional)

Our application fails the anti-tamper self-check if Gradle 8.1 or 8.1.1 is used.

### Context

It appears that Gradle 8.1 and 8.1.1 are automatically altering the bytecode and/or changing the class initialization sequence resulting in the application crashing.

Setting the org.gradle.internal.instrumentation.agent=false environment variable has made no difference. 

**The feature requested is that there should be a way to tell Gradle to not alter the application bytecode in ANY way.**

**Comment by jasonsparc** [comment]
Hi! I just wanted to say that, my issue might be related to your issue. See, https://github.com/gradle/gradle/issues/25164

**Comment by zelix0** [comment]
With further testing we have found that the first Gradle version to trigger the anti-tamper fail was Gradle 6.6. The SHA256 hashes of some classes are changed by Gradle 6.6.  Gradle 8.1 and 8.1.1 make extra changes possibly to the class initialization order.

**Comment by zelix0** [comment]
Have done further testing using Gradle 6.6. (The application fails too abruptly under Gradle 8.1 to allow useful debugging.) When comparing the original application JAR with the JAR in the Gradle cache it is clear that Gradle 6.6 is 

1. Reordering the constant pool
2. Adding constant pool entries and
3. Inserting calls to org/gradle/internal/classpath/Instrumented.systemProperty:(String,String,String) 

I have set org.gradle.internal.instrumentation.agent=false as an environment variable and in gradle.properties but it has made no difference.  There needs to be a way to tell Gradle to not tamper with the classes it is running in ANY way. Otherwise application which do an anti-tamper check or which are otherwise sensitive to alteration will fail under Gradle 6.6 or better.

**Comment by zelix0** [comment]
[qlq_original.txt](https://github.com/gradle/gradle/files/11580398/qlq_original.txt)
[qlq_cached.txt](https://github.com/gradle/gradle/files/11580399/qlq_cached.txt)

I have attached the javap output for the class qlp.class.  The file qlq_original.txt shows the class contents of the original qlq class that was executed through Gradle 6.6.  The file qlq_cached.txt shows the contents of the class retrieved from the Gradle cache.  It is this cached version which Gradle actually executed and has modified.

As can be seen, Gradle 6.6 has clearly modified the class.  Amongst other things Gradle has inserted a call to org/gradle/internal/classpath/Instrumented.systemProperty:(String,String,String).

Note that for any application which does a SHA anti-tampering self check ANY change is fatal.  There needs to be way of switching off this behavior.

**Comment by zelix0** [comment]
[testCase0.zip](https://github.com/gradle/gradle/files/11581134/testCase0.zip)

I have attached an absolutely minimal test case which demonstrates that Gradle 8.1.1 makes changes to an executed class. Please see the instructions in the contained ReadMe.md.  

The changes made by Gradle are not be fatal in this simplified case but they would be if there was an anti-tampering self-check.

Note that for any application which does a SHA anti-tampering self check ANY change is fatal. **There needs to be way of switching off this behavior.**  Otherwise such application will not be able to run through any version of Gradle after 6.5.1.

**Comment by eskatos** [comment]
Thanks! This clarifies things a bit.

Your test case shows the circumstances of your observation but doesn't allow reproduction of the problem because `com.zelix.gradle:plugin:1.0.0` is not available. I understand this is closed source.

As stated in the above [comment](https://github.com/gradle/gradle/issues/25159#issuecomment-1564351040):

> Gradle does instrumentation of classes on the build script classpath, but not on your production code.

IOW Gradle instruments classes of Gradle plugins, not of applications. In your case the affected classes are indeed from a Gradle plugin. I edited the issue title to reflect that instead of implying that Gradle instruments application code.

This instrumentation allows for Gradle features such as the automatic detection of build logic inputs for [Configuration Cache](https://docs.gradle.org/current/userguide/configuration_cache.html) invalidation.

What would work in your case then would be a way to opt-out of the plugin classpath instrumentation. Such an opt-out would defeat Gradle features that rely on it. We could imagine disabling those features if you opt-out of the instrumentation.

----

Here is the description of a workaround. Here I'm assuming your Gradle plugin is constitued of some integration with Gradle  and of Gradle task(s) that make use of your closed source product to process the application code that is built.

One approach that would work Today would be to split your Gradle plugin into two parts:
1. a thin layer for the Gradle integration
2. the actual implementation of your tasks actions

The former would be instrumented by Gradle and would not have an anti-tamper check.
The latter would be resolved as a separate JAR and could keep an anti-tamper check.

The above can be achieved by having the Gradle plugin declare a dependency to the separate JAR in a [Configuration](https://docs.gradle.org/current/userguide/declaring_dependencies.html#sec:what-are-dependency-configurations) created by the plugin.

Then the Gradle plugin can register task(s) that make use of the [Worker API](https://docs.gradle.org/current/userguide/worker_api.html). With the Worker API, tasks can execute code from external JARs in [isolation](https://docs.gradle.org/current/userguide/worker_api.html#changing_the_isolation_mode). These JARs are not instrumented and can have an anti-tamper check.

I understand this requires work on your Gradle plugin and would move that thin layer outside of the protection provided by your product but it is the only option without a potential opt-out of Gradle plugins instrumentation.

----

I'll let our relevant team take over this issue and make a decision about the feasibility of the opt-out.

**Comment by zelix0** [comment]
> com.zelix.gradle:plugin:1.0.0 is not available

Sorry. I put the wrong download link in the ReadMe.md.  The correct link is https://www.zelix.com/klassmaster/download/zkm-gradle-plugin-1.0.0.zip.

> Gradle does instrumentation of classes on the build script classpath, but not on your production code

The plugin basically collects parameters from the user and then uses them to run a local JAR which is closed source. The plugin is merely a means for allowing the closed source local JAR to be run using Gradle. The essence of the problem is that the classes of this local JAR are being modified causing them to fail the anti-tamper self-check.

**Is it possible to exclude local JAR files from such modifications?** They were never intended to be treated as part of the plugin.

**Comment by zelix0** [comment]
> One approach that would work Today would be to split your Gradle plugin into two parts

This exactly what we thought we had done.  The plugin references a separate local JAR but this local JAR is being modified.

> edited the issue title to reflect that instead of implying that Gradle instruments application code

I am not sure that the change in the title was appropriate.  The separate local JAR is the application that the plugin initiates.

**Comment by eskatos** [comment]
Gradle instruments the "plugin classpath" and your second jar is currently part of that classpath. The split needs to move that other JAR outside of the "plugin classpath". In other words for the other JAR not to be instrumented it should not be a dependency of the plugin but be declared in a separate configuration . If you follow my suggestion to use the Worker API in the task that runs that external JAR by resolving it from a separate configuration it should work.

**Comment by zelix0** [comment]
I changed the title back to "Gradle 8.1 and 8.1.1 alter Gradle **application** bytecode so that it fails anti-tamper check" to avoid the risk of misunderstanding. The plugin itself is downloaded from https://www.zelix.com/klassmaster/download/zkm-gradle-plugin-1.0.0.zip.  It contains no local closed source application JAR file.

The test case https://github.com/gradle/gradle/files/11581134/testCase0.zip includes the local JAR ZKM.jar which simulates the closed source application JAR which is totally independent of the plugin.  The plugin merely allows the ZKM.jar application JAR to be run from within Gradle.  

It is this local ZKM.jar application JAR which is being inappropriately modified by Gradle (after version 6.5.1) resulting in the failure of the anti-tampering self-check.

**Comment by zelix0** [comment]
> If you follow my suggestion to use the Worker API in the task that runs that external JAR by resolving it from a separate configuration it should work

To be honest I couldn't make sense of the Worker API example. Of course it was focused on getting a task to operate in parallel which is not the issue here.  

Shouldn't setting the org.gradle.internal.instrumentation.agent=false environment variable be enough to prevent instrumentation?  In the case of the test case, the changes that are made to the ZKM.class appear to quite gratuitous. The constant pool is reordered for no apparent reason.

> Gradle instruments the "plugin classpath"...

But does it really have to do this in the case of local JAR files that are not in any formal repository?  Note that this issue #25159 is a "feature request".

**Comment by zelix0** [comment]
Note that because there is apparently no way of telling the newer versions of Gradle to not alter the local JAR classes in any way we are having to tell our users to use nothing newer than Gradle 6.5.1. This is not good for anyone.

**Comment by mlopatkin** [comment]
tl;dr Worker API with classpath/process isolation is the currently supported way of excluding the JARs from bytecode instrumentation. More details follow.

Gradle instruments the buildscript classpath (compiled build scripts, plugins and their runtime dependencies) to track various external inputs to the build configuration (mostly to support configuration caching), ensure task cacheability, and to provide some backward compatibility features. There's no way for Gradle to figure out how parts of the classpath are used, and excluding parts of the classpath from instrumentation opens the room for broken builds (build may fail or may silently re-use cached state when it shouldn't).

However, the Worker API provides the necessary level of isolation - code run in the isolated worker cannot affect the build configuration/cacheability, so Gradle doesn't instrument it. To provide the isolation the code is loaded in the separate classloader (or even in the separate process) with separate classpath. The limitation is that it isn't possible to communicate anything back to the build script code from the worker except the failure, but looking at your plugin - it isn't a problem.

Another option could be to use `JavaExec` task which also uses a separate classpath (and a separate process, of course), but this would likely be less convenient (your JAR has to have the `main()` entry point). The classpath of `JavaExec` is not instrumented, of course.

>To be honest I couldn't make sense of the Worker API example. Of course it was focused on getting a task to operate in parallel which is not the issue here.

The only difference is that for parallelism you have to submit multiple `WorkerAction`s to the `WorkQueue`, everything else still holds. 
1. Make sure your `compile file("...")` is a `compileOnly` instead, so it doesn't become a runtime dependency of the plugin itself:
```gradle
dependencies {
    compileOnly gradleApi()
    compileOnly files("$ZKM_JAR_PATH") //ZKM_JAR_PATH can be set in gradle.properties
}
```
2. Extract code that uses classes from the JAR to the `WorkerAction` implementation. The action implementation is going to be loaded in the Worker's class loader, and won't be instrumented, as well as the rest of the worker classpath. For your class it could be like:
```java
public abstract class ZKMAction implements WorkAction<ZKMAction.Params> {
    interface Params extends WorkParameters {
        Property<String> getScriptFile();
        Property<String> getLogFileName();
        Property<String> getTrimLogFileName();
        Property<String> getDefaultExcludeFileName();
        Property<String> getDefaultTrimExcludeFileName();
        Property<String> getDefaultDirectoryName();
        Property<Boolean> getVersbose();
        Property<Boolean> getParseOnly();
        MapProperty<Object, Object> getExtraProperties();
    }

    @Override
    public void execute() {
        Params params = getParameters();

        com.zelix.ZKM.run(
                params.getScriptFile().getOrNull(),
                params.getLogFileName().getOrNull(),
                params.getTrimLogFileName().getOrNull(),
                params.getDefaultExcludeFileName().getOrNull(),
                params.getDefaultTrimExcludeFileName().getOrNull(),
                params.getDefaultDirectoryName().getOrNull(),
                params.getVersbose().getOrNull(),
                params.getParseOnly().getOrNull(),
                params.getExtraProperties().getOrNull());
    }
}
```
3. Build a configuration to be used as worker's classpath. The example uses `ConfigurableFileCollection` field of task for this that is initialized in the build script by the user. Alternatively, you can define the configuration and add the JAR to it in the plugin code, then wire it to the task. If the path to JAR comes from the `gradle.properties`, you can do something like:
```java
public abstract class ZKMPluginTask extends DefaultTask {

   @InputFiles
   public abstract ConfigurableFileCollection getClasspath();
// ... snip
}

public class ZKMPlugin implements Plugin<Project> {
    @Override
    public void apply(Project project) {
        project.getExtensions().create("zkmSetting", ZKMPluginExtension.class);
        project.getTasks().create("zkmTask", ZKMPluginTask.class, task -> {
            task.getClasspath().from(project.file(project.property("ZKM_JAR_PATH")));
        });
    }
}
```
4. Build a `WorkQueue` out of this configuration in the task action, then configure and submit the action itself:
```java
public abstract class ZKMPluginTask extends DefaultTask {
// snip
   @Inject
   protected abstract WorkerExecutor getWorkerExecutor();
// snip
   @TaskAction
   public void run() {
    // getting stuff from the extension skipped
    getWorkerExecutor().classLoaderIsolation(clSpec -> {
         clSpec.getClasspath().setFrom(getClasspath());
      }).submit(ZKMAction.class, params -> {
         params.getScriptFile().set(scriptFile.getAbsolutePath());
         params.getLogFileName().set(logFileName);
         params.getTrimLogFileName().set(trimLogFileName);
         params.getDefaultExcludeFileName().set(defaultExcludeFileName);
         params.getDefaultTrimExcludeFileName().set(defaultTrimExcludeFileName);
         params.getDefaultDirectoryName().set(defaultDirectoryName);
         params.getVersbose().set(isVerbose);
         params.getParseOnly().set(isParseOnly);
         params.getExtraProperties().set(extraProperties);
      });
}
```
You can also look at the sources without snips at https://gist.github.com/mlopatkin/dc36350e0589df782f7f2c4cc41cbf1c

>Shouldn't setting the org.gradle.internal.instrumentation.agent=false environment variable be enough to prevent instrumentation?

No, it doesn't disable instrumentation, just switches to the legacy way of doing it. In general, I advise not to use `internal` switches, as these can go away without notice, and there's no plan to support this particular one longer term.

>which simulates the closed source application JAR which is totally independent of the plugin. The plugin merely allows the ZKM.jar application JAR to be run from within Gradle.

You're adding this JAR as an implementation dependency of your plugin, so Gradle has no way of knowing that the JAR is "totally independent" - there's no difference to it being just a library, so Gradle assumes that it can affect the build configuration.

>But does it really have to do this in the case of local JAR files that are not in any true repository? 

It doesn't really matter where the JAR comes from, as all have the same capabilities when used in the buildscript classpath.

>I changed the title back to "Gradle 8.1 and 8.1.1 alter Gradle application bytecode so that it fails anti-tamper check" to avoid the risk of misunderstanding.

"application" heavily implies that Gradle instruments the code being built with Gradle which is actually misleading. It is a big stretch to call anything on the buildscript classpath an "application".

PS. There is a room for other improvements in plugin, like supporting incremental builds and configuration cache, but this is out of scope.

**Comment by zelix0** [comment]
Thank you Mikhail for all the work you have put in above.  I understand that it is intended to be a complete working example of the Worker API approach.  I have attempted to put it into a project structure in the attached ZKMAction1.zip.  However, when I run "gradlew build" I get the Java compile error below.  The "missing" class com.zelix.ZKM is in the ZKM.jar which is pointed to by the ZKM_JAR_PATH property that is set in the gradle.properties. However, when I use the --debug option I can see that the classpath being passed to the Java compiler does not include the ZKM.jar.  I have tried using absolute & relative paths for the ZKM_JAR_PATH property but it makes no difference.  Is there something I am missing?

D:\Projects\gradle\ZKMAction\buildSrc\src\main\java\com\zelix\gradle\ZKMAction.java:25: error: cannot find symbol
         com.zelix.ZKM.run(
   symbol:   class ZKM
   location: package com.zelix



[ZKMAction1.zip](https://github.com/gradle/gradle/files/11670744/ZKMAction1.zip)

**Comment by eskatos** [comment]
Hi,

In your zip, the plugin is in `buildSrc` so the `compileOnly` dependency to the jar must be set in `buildSrc/build.gradle`. Moreover, the plugin was not applied in the root `build.gradle` script.

Please find here a modified ZIP that works AFAICT:

[ZKMAction.zip](https://github.com/gradle/gradle/files/11744621/ZKMAction.zip)

**Comment by zelix0** [comment]
Thank you @eskatos for your changes and clarifications. The project is now building successfully.  The next step in the testing process is to install the plugin to the local .m2 repository.  With earlier versions of Gradle we simply ran "gradlew build install" and all was simple & straightforward.  Now with the later versions of Gradle "install" no longer works.  My understanding is that we should now use publishToMavenLocal.  However, when we do this we get a ":publishToMavenLocal UP-TO-DATE" response but nothing goes into the local repository.

Is this because of the more complicated project structure or is it something else?  I apologize for our lack of Gradle expertise. After spending hours trying to understand the Gradle documentation and attempting to replicate possibly obsolete approaches suggested on StackOverflow I have to (at least temporarily) admit defeat.

I Have attached our (slightly changed) project structure as it currently is.
[ZKMAction2.zip](https://github.com/gradle/gradle/files/11765716/ZKMAction2.zip)

**Comment by mlopatkin** [comment]
> Is this because of the more complicated project structure or is it something else? I apologize for our lack of Gradle expertise. After spending hours trying to understand the Gradle documentation and attempting to replicate possibly obsolete approaches suggested on StackOverflow I have to (at least temporarily) admit defeat.

I'm sorry, but I don't think this issue is a good medium to get help about how to use Gradle, the discussion is better suited for [Gradle forums](https://discuss.gradle.org/) or the [community Slack](https://gradle.org/slack-invite).

I've updated [the documentation](https://docs.gradle.org/nightly/userguide/configuration_cache.html#config_cache:requirements:bytecode_modifications_and_isolation) to highlight the fact that Gradle instruments bytecode and provide links to the Worker API mitigation.
